File: javascript/0092-reverse-linked-list-ii.js
var reverseBetween = function(head, left, right) {
    
  let dummy = new ListNode(null, head);
  let preNode = dummy;
  let current = head;
    let count = 0;
    while(count < left - 1) {
        current = current.next;
        preNode = preNode.next;
        count++;
    }
    
    count = 0;
    let pre = null;
    while(count < right - left + 1) {
        const temp = current.next;
        current.next = pre;
        pre = current;
        current = temp;
        count++;
    }
    
    preNode.next.next = current;
    preNode.next = pre;
    return dummy.next;
};

File: kotlin/0092-reverse-linked-list-ii.kt
class Solution {
    fun reverseBetween(head: ListNode?, left: Int, right: Int): ListNode? {
        val dummy: ListNode? = ListNode(0)
        dummy?.next = head

        var cur = dummy
        repeat (left - 1) {
            cur = cur?.next
        }

        var pre = cur
        var start = cur?.next
        var end = start?.next

        repeat (right - left) {
            start?.next = end?.next
            end?.next = pre?.next
            pre?.next = end
            end = start?.next
        }
        
        return dummy?.next
    }
}
